### PR TITLE
PoC: config reading with state v2

### PR DESCRIPTION
**Pull request by poslegm**
My next attempt to implement config reading with state. Based on #88 and @kitbellew's `ConfigDecoderFactory` idea.

Now it presents independent `ConfigReaderDecoder` typeclass with main API
```scala
def decoder: S => ConfDecoder[A]
```

where `S` is state provided on decoder construction (like default values for collections).

It still has priority implicits magic but with improvements:

1. safe methods for map/orElse/etc on `ConfDecoderReader` without state loss
2. doesn't break `ConfDecoder` API and already created decoders
3. No inheritance from `ConfDecoder`

I defined `ConfDecoderReader` instances for `Map[String, A]` and `Iterable[A]` with ability to `.add` to defaults. It's possible to implement `.remove` also.

`S` type parameter allows users to provide rich states to custom decoders. If you think this is an unnecessary overcomplication, I don't mind replacing it with default only. 

What do you think?

### CODE REVIEW

File: metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala [hunk 1]
```diff
@@ -33,7 +34,7 @@ sealed abstract class Conf extends Product with Serializable {
     ConfGet.get(this, path, extraNames: _*)
   def getOrElse[T](path: String, extraNames: String*)(
       default: T
-  )(implicit ev: ConfDecoder[T]): Configured[T] =
+  )(implicit ev: ConfDecoderWithDefaultMaybe[T]): Configured[T] =
     ConfGet.getOrElse(this, default, path, extraNames: _*)
```
Comment (kitbellew) [inline metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala:38]: do you need to change the implicit for this call as well? will `ConfGet.getOrElse` use the `default` parameter in its first matching case (the one currently not using default)?

File: metaconfig-tests/shared/src/test/scala/metaconfig/ConfListSuite.scala [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      anotherField: List[String] = Nil,
+      bar: Bar = Bar()
+  )
+  implicit val surface: generic.Surface[Foo] =
+    generic.deriveSurface
+
+  implicit val reader: ConfDecoder[Foo] =
+    generic.deriveDecoder[Foo](Foo()).noTypos
+
+  case class Nested(raw: String = "x")
```
Comment (kitbellew) [inline metaconfig-tests/shared/src/test/scala/metaconfig/ConfListSuite.scala:24]: there are multi-nested fields in `AllTheAnnotations`, can we test deeper?

Comment (poslegm) [reply metaconfig-tests/shared/src/test/scala/metaconfig/ConfListSuite.scala:24]: Added test with multiple nesting

File: metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala (hunk 1 above)
Comment (poslegm) [reply metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala:38]: Hm, I don't found usage of this method in macros. But I would leave the defaults support here for uniformity with `getSettingOrElse` method which has same functionality

Comment (kitbellew) [reply metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala:38]: i meant `ConfGet.getOrElse`, it's used by `getSettingsOrElse` (which is used in macros).

**Comment by poslegm** [comment]
> can you clarify: are you solving just the .add case for lists (or possibly maps), or will this also work for scalafmt's presets?

My main task here is the `.add` support. I think that generic state propagation mechanic can help with delivering external input data (like presets) to custom decoders. But I don't play with it yet and I can't guarantee that.

File: metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala (hunk 1 above)
Comment (poslegm) [reply metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala:38]: Then I don't understand the question. `ConfGet.getOrElse` [uses](https://github.com/scalameta/metaconfig/pull/92/files#diff-9aff91f762e9b2b1da37ce46226c0373R24-R25) default parameter in its first matching case

Comment (kitbellew) [reply metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala:38]: the question is: you changed `ev` for `Conf.getOrElse` and `Conf.getSettingOrElse` to be `ConfDecoderWithDefaultMaybe`; but _i think_ you didn't make the same change in the only method they call (`ConfGet.getOrElse`). so either you don't need `WithDefaultMaybe` in this file, or you *do* need it in `ConfGet`.

File: metaconfig-core/shared/src/main/scala/metaconfig/internal/ConfGet.scala [hunk 3]
```diff
@@ -20,10 +18,11 @@ object ConfGet {
     }
 
   def getOrElse[T](conf: Conf, default: T, path: String, extraNames: String*)(
-      implicit ev: ConfDecoder[T]
+      implicit ev: ConfDecoderWithDefaultMaybe[T]
```
Comment (kitbellew) [inline metaconfig-core/shared/src/main/scala/metaconfig/internal/ConfGet.scala:21]: sorry. i was blind. i don't know why. ignore the last comment. brrr...

Comment (poslegm) [reply metaconfig-core/shared/src/main/scala/metaconfig/internal/ConfGet.scala:21]: :ok_hand:

**Comment by olafurpg** [comment]
If there is no rush on your end to get this functionality merged, would it be OK if you give me some time to try and draft an alternative solution that avoids breaking changes and also unblocks other use-cases besides defaults?

I'm not convinced it's the best way forward to use the design in this PR. I would love to have a solution that is more general that doesn't require the same advanced implicit machinery.

**Comment by poslegm** [comment]
Yes, of course!

**Comment by kitbellew** [comment]
> If there is no rush on your end to get this functionality merged, would it be OK if you give me some time to try and draft an alternative solution that avoids breaking changes and also unblocks other use-cases besides defaults?
> 
> I'm not convinced it's the best way forward to use the design in this PR. I would love to have a solution that is more general that doesn't require the same advanced implicit machinery.

@olafurpg are you still thinking about an alternative? i saw a comment about your focus in `moped` now. any recommendation what we should do to support the functionality that @poslegm originally wanted to implement?

**Pull request closed without merging**